### PR TITLE
Primewire

### PR DESCRIPTION
**Pull request by zisra**
**Changes:**
- Add primewire source
- Add voe embed
- Add streamtape embed
- Fix dood embed

### CODE REVIEW

File: src/providers/sources/primewire/blowfish.ts [hunk 1]
```diff
@@ -0,0 +1,378 @@
+const pArray = [
```
Comment (JorrinKievit) [inline src/providers/sources/primewire/blowfish.ts:1]: Can we move these arrays to its own file?

File: src/providers/sources/primewire/blowfish.ts
Comment (JorrinKievit) [inline src/providers/sources/primewire/blowfish.ts:1]: Could we have a short description on what this file is, how to maintain, etc?

File: src/providers/sources/primewire/blowfish.ts [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+        const secondCharCode = input.charCodeAt(i + 1);
+        const thirdCharCode = input.charCodeAt(i + 2);
+        encoded += String.fromCharCode(((15 & charCode) << 12) | ((63 & secondCharCode) << 6) | (63 & thirdCharCode));
+        i += 2;
+      }
+    }
+    return encoded;
+  }
+
+  bd(e: string) {
```
Comment (JorrinKievit) [inline src/providers/sources/primewire/blowfish.ts:341]: Rename this function to something descriptive

File: src/providers/sources/primewire/index.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  const titlePage = load(title);
+  const userData = titlePage('#user-data').attr('v');
+  if (!userData) throw new NotFoundError('No user data found');
+
+  const links = getLinks(userData);
+
+  const embeds = [];
+
+  for (const link in links) {
+    if (link.includes(link)) {
```
Comment (JorrinKievit) [inline src/providers/sources/primewire/index.ts:35]: This always resolves to true?

File: src/providers/sources/primewire/index.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+          break;
+        case 'streamvid.net':
+          embedId = 'streamvid';
+          break;
+        default:
+          embedId = null;
+      }
+      if (!embedId) continue;
+      embeds.push({
+        url: `https://www.primewire.tf/links/go/${links[link]}`,
```
Comment (JorrinKievit) [inline src/providers/sources/primewire/index.ts:57]: can use the baseUrl here

File: src/providers/sources/primewire/blowfish.ts [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+      }
+      if (a !== 64) {
+        s += String.fromCharCode(r);
+      }
+    }
+    return s;
+  }
+}
+
+export function getLinks(e: string) {
```
Comment (JorrinKievit) [inline src/providers/sources/primewire/blowfish.ts:367]: What is e?

File: src/providers/sources/primewire/blowfish.ts [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+  const t = e.slice(-10);
```
Comment (JorrinKievit) [inline src/providers/sources/primewire/blowfish.ts:368]: Same for these, a descriptive variable name if possible

File: src/providers/sources/primewire/index.ts (hunk 3 above)
Comment (zisra) [reply src/providers/sources/primewire/index.ts:35]: https://eslint.org/docs/latest/rules/guard-for-in

File: src/providers/sources/primewire/index.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+export const primewireScraper = makeSourcerer({
+  id: 'primewire',
+  name: 'Primewire',
+  rank: 350,
+  flags: [flags.CORS_ALLOWED],
+  async scrapeMovie(ctx) {
+    if (!ctx.media.imdbId) throw new Error('No imdbId provided');
+    const searchResult = await search(ctx, ctx.media.imdbId);
+
+    const title = await ctx.fetcher<string>(`movie/${searchResult}`, {
```
Comment (JorrinKievit) [inline src/providers/sources/primewire/index.ts:80]: This should be proxied

File: src/providers/sources/primewire/index.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+    return {
+      embeds,
+    };
+  },
+  async scrapeShow(ctx) {
+    if (!ctx.media.imdbId) throw new Error('No imdbId provided');
+    const searchResult = await search(ctx, ctx.media.imdbId);
+
+    const season = await ctx.fetcher<string>(`tv/${searchResult}`, {
```
Comment (JorrinKievit) [inline src/providers/sources/primewire/index.ts:94]: This too

**Review by JorrinKievit — CHANGES_REQUESTED** [review]
Few remarks:

- Upstream returns a 403 for the HLS playlist. Probably needs a referer so should be used extension only (IP_LOCKED flag)?
- The streams are not that fast, are mostly 720p and have no subtitles. So for that reason I would rank it below VidSrcTo (I'm gonna rerank for the new release anyway but just for this PR's sake)
- Could be worth investigating if its faster on primewire's site itself, and we can add headers to the streams for faster load speeds?

**Review by JorrinKievit — APPROVED** [review]

**Pull request merged**